contrib/bsp/src/mill/contrib/BSP.scala

```diff
@@ -31,107 +31,126 @@ object BSP extends ExternalModule {
-    * until a client connects and ends the connection
-    * after the client sent an "exit" notification
-    *
+   * Computes a mill command which starts the mill-bsp
+   * server and establishes connection to client. Waits
+   * until a client connects and ends the connection
+   * after the client sent an "exit" notification
+   *
     * @param ev Environment, used by mill to evaluate commands
-    * @return: mill.Command which executes the starting of the
-    *          server
-    */
-  def start(ev: Evaluator): Command[Unit] = T.command {
-    val eval = new Evaluator(ev.home, ev.outPath, ev.externalOutPath, ev.rootModule, ev.baseLogger, ev.classLoaderSig,
-                             ev.workerCache, ev.env, false)
-    val millServer = new mill.contrib.bsp.MillBuildServer(eval, bspProtocolVersion, version, languages)
-    val executor = Executors.newCachedThreadPool()
+   * @return: mill.Command which executes the starting of the
+   *          server
+   */
+  def start(ev: Evaluator): Command[Unit] =
+    T.command {
+      val eval = new Evaluator(
+        ev.home,
+        ev.outPath,
+        ev.externalOutPath,
+        ev.rootModule,
+        ev.baseLogger,
+        ev.classLoaderSig,
+        ev.workerCache,
+        ev.env,
+        false
+      )
+      val millServer = new MillBuildServer(eval, bspProtocolVersion, BuildInfo.millVersion)
+      val executor = Executors.newCachedThreadPool()
 
-    val stdin = System.in
-    val stdout = System.out
-    try {
-      val launcher = new Launcher.Builder[BuildClient]()
-        .setOutput(stdout)
-        .setInput(stdin)
-        .setLocalService(millServer)
-        .setRemoteInterface(classOf[BuildClient]).
-        traceMessages(new PrintWriter((os.pwd / "bsp.log").toIO))
-        .setExecutorService(executor)
-        .create()
-      millServer.onConnectWithClient(launcher.getRemoteProxy)
-      val listening = launcher.startListening()
-      millServer.cancelator = () => listening.cancel(true)
-      val voidFuture = listening.get()
-    } catch {
-      case _: CancellationException => System.err.println("The mill server was shut down.")
-      case e: Exception =>
-        System.err.println("An exception occured while connecting to the client.")
-        System.err.println("Cause: " + e.getCause)
-        System.err.println("Message: " + e.getMessage)
-        System.err.println("Exception class: " + e.getClass)
-        System.err.println("Stack Trace: " + e.getStackTrace)
-    } finally {
-      System.err.println("Shutting down executor")
-      executor.shutdown()
+      val stdin = System.in
+      val stdout = System.out
+      try {
+        val launcher = new Launcher.Builder[BuildClient]()
+          .setOutput(stdout)
+          .setInput(stdin)
+          .setLocalService(millServer)
+          .setRemoteInterface(classOf[BuildClient])
+          .traceMessages(new PrintWriter((os.pwd / ".bsp" / "mill.log").toIO))
```

Comment:
I moved the log file in the `.bsp` folder.
I think it was too much in the face of the user.